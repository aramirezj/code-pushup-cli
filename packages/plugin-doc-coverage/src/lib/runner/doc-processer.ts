import { ClassDeclaration, Project, SourceFile } from 'ts-morph';
import type { DocCoveragePluginConfig } from '../config.js';
import type {
  CoverageResult,
  CoverageType,
  UnprocessedCoverageResult,
} from './models.js';
import {
  calculateCoverage,
  createEmptyUnprocessedCoverageReport,
  getCoverageTypeFromKind,
} from './utils.js';

/**
 * Processes documentation coverage for TypeScript files in the specified path
 * @param toInclude - The file path pattern to include for documentation analysis
 * @returns {CoverageResult} Object containing coverage statistics and undocumented items
 */
export function processDocCoverage(
  config: DocCoveragePluginConfig,
): CoverageResult {
  const project = new Project();
  project.addSourceFilesAtPaths(config.sourceGlob);
  return getUnprocessedCoverageReport(project.getSourceFiles());
}

/**
 * Gets the unprocessed coverage report from the source files
 * @param sourceFiles - The source files to process
 * @returns {UnprocessedCoverageResult} The unprocessed coverage report
 */
export function getUnprocessedCoverageReport(sourceFiles: SourceFile[]) {
  const unprocessedCoverageReport = sourceFiles.reduce(
    (coverageReportOfAllFiles, sourceFile) => {
      // Info of the file
      const filePath = sourceFile.getFilePath();
      const classes = sourceFile.getClasses();

      // All nodes of the file
      const allNodesFromFile = [
        ...sourceFile.getFunctions(),
        ...classes,
        ...getClassNodes(classes),
        ...sourceFile.getTypeAliases(),
        ...sourceFile.getEnums(),
        ...sourceFile.getInterfaces(),
        // ...sourceFile.getVariableStatements().flatMap(statement => statement.getDeclarations())
      ];

      const coverageReportOfCurrentFile = allNodesFromFile.reduce(
        (acc, node) => {
          const nodeType = getCoverageTypeFromKind(node.getKind());
          acc[nodeType].nodesCount++;
          if (node.getJsDocs().length === 0) {
            acc[nodeType].issues.push({
              file: filePath,
              type: nodeType,
              name: node.getName() || '',
              line: node.getStartLineNumber(),
            });
          }
          return acc;
        },
        createEmptyUnprocessedCoverageReport(),
      );

      return mergeCoverageResults(
        coverageReportOfAllFiles,
        coverageReportOfCurrentFile,
      );
    },
    createEmptyUnprocessedCoverageReport(),
  );

  return calculateCoverage(unprocessedCoverageReport);
}

/**
 * Merges two coverage results
 * @param results - The first empty coverage result
 * @param current - The second coverage result
 * @returns {UnprocessedCoverageResult} The merged coverage result
 */
export function mergeCoverageResults(
  results: UnprocessedCoverageResult,
  current: Partial<UnprocessedCoverageResult>,
) {
  return {
    ...Object.fromEntries(
      Object.entries(results).map(([key, value]) => {
        const node = value as CoverageResult[CoverageType];
        const type = key as CoverageType;
        return [
          type,
          {
            nodesCount: node.nodesCount + (current[type]?.nodesCount ?? 0),
            issues: [...node.issues, ...(current[type]?.issues ?? [])],
          },
        ];
      }),
    ),
  } as UnprocessedCoverageResult;
}

/**
 * Gets the nodes from a class
 * @param classNodes - The class nodes to process
 * @returns {Node[]} The nodes from the class
 */
export function getClassNodes(classNodes: ClassDeclaration[]) {
  return classNodes.flatMap(classNode => [
    ...classNode.getMethods(),
    ...classNode.getProperties(),
  ]);
}
