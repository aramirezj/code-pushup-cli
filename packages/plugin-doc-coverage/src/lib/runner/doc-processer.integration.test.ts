import { processDocCoverage } from './doc-processer';

describe('processDocCoverage', () => {
  it('should succesfully get the right number of ts files', () => {
    const results = processDocCoverage({
      sourceGlob: [
        'packages/plugin-doc-coverage/mocks/fixtures/angular/**/*.ts',
      ],
    });
    expect(results).toMatchSnapshot();
  });

  it('should succesfully get the right number of ts files and not include spec files', () => {
    const results = processDocCoverage({
      sourceGlob: [
        'packages/plugin-doc-coverage/mocks/fixtures/angular/**/*.ts',
        '!**/*.spec.ts',
        '!**/*.test.ts',
      ],
    });
    expect(results).toMatchSnapshot();
  });
});
