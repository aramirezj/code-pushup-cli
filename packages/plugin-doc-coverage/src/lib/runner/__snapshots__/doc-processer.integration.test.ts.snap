// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processDocCoverage > should succesfully get the right number of ts files 1`] = `
{
  "classes": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 1,
  },
  "enums": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "functions": {
    "coverage": 33.333333333333336,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/app.component.spec.ts",
        "line": 1,
        "name": "notRealisticFunction",
        "type": "functions",
      },
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/map-event.function.ts",
        "line": 1,
        "name": "mapEventToCustomEvent",
        "type": "functions",
      },
    ],
    "nodesCount": 3,
  },
  "interfaces": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "methods": {
    "coverage": 50,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/app.component.ts",
        "line": 15,
        "name": "sendEvent",
        "type": "methods",
      },
    ],
    "nodesCount": 2,
  },
  "properties": {
    "coverage": 0,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/app.component.ts",
        "line": 5,
        "name": "title",
        "type": "properties",
      },
    ],
    "nodesCount": 1,
  },
  "types": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "variables": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
}
`;

exports[`processDocCoverage > should succesfully get the right number of ts files and not include spec files 1`] = `
{
  "classes": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 1,
  },
  "enums": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "functions": {
    "coverage": 50,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/map-event.function.ts",
        "line": 1,
        "name": "mapEventToCustomEvent",
        "type": "functions",
      },
    ],
    "nodesCount": 2,
  },
  "interfaces": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "methods": {
    "coverage": 50,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/app.component.ts",
        "line": 15,
        "name": "sendEvent",
        "type": "methods",
      },
    ],
    "nodesCount": 2,
  },
  "properties": {
    "coverage": 0,
    "issues": [
      {
        "file": "/home/alejandro/dev/code-pushup-cli/packages/plugin-doc-coverage/mocks/fixtures/angular/app.component.ts",
        "line": 5,
        "name": "title",
        "type": "properties",
      },
    ],
    "nodesCount": 1,
  },
  "types": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
  "variables": {
    "coverage": 100,
    "issues": [],
    "nodesCount": 0,
  },
}
`;
